x/evm/keeper/msg_server.go

```diff
@@ -0,0 +1,187 @@
+	return core.TransactionToMessage(tx, signer, server.getBaseFee())
+}
+
+func (server msgServer) applyEVMMessage(ctx sdk.Context, msg *core.Message, stateDB vm.StateDB, gp core.GasPool) (*core.ExecutionResult, error) {
+	coinbase, err := server.GetFeeCollectorAddress(ctx)
+	if err != nil {
+		return nil, err
+	}
+	cfg := server.GetChainConfig(ctx).EthereumConfig(server.ChainID())
+	blockCtx := vm.BlockContext{
+		CanTransfer: core.CanTransfer,
+		Transfer:    core.Transfer,
+		GetHash:     server.getHashFn(ctx),
+		Coinbase:    coinbase,
+		GasLimit:    gp.Gas(),
+		BlockNumber: big.NewInt(ctx.BlockHeight()),
+		Time:        uint64(ctx.BlockHeader().Time.Unix()),
+		Difficulty:  big.NewInt(0),       // only needed for PoW
+		BaseFee:     server.getBaseFee(), // feemarket not enabled
+		Random:      nil,                 // not supported
+	}
+	txCtx := core.NewEVMTxContext(msg)
+	evmInstance := vm.NewEVM(blockCtx, txCtx, stateDB, cfg, vm.Config{})
+	st := core.NewStateTransition(evmInstance, msg, &gp)
+	return st.TransitionDb()
+}
+
+func (server msgServer) writeReceipt(ctx sdk.Context, tx *ethtypes.Transaction, msg *core.Message, res *core.ExecutionResult) error {
+	var contractAddr common.Address
+	if msg.To == nil {
+		contractAddr = crypto.CreateAddress(msg.From, msg.Nonce)
+	} else if len(msg.Data) > 0 {
+		contractAddr = *msg.To
+	}
+
+	cumulativeGasUsed := res.UsedGas
+	if ctx.BlockGasMeter() != nil {
+		limit := ctx.BlockGasMeter().Limit()
+		cumulativeGasUsed += ctx.BlockGasMeter().GasConsumed()
+		if cumulativeGasUsed > limit {
+			cumulativeGasUsed = limit
+		}
+	}
+
+	receipt := &types.Receipt{
+		TxType:            uint32(tx.Type()),
+		CumulativeGasUsed: cumulativeGasUsed,
+		TxHashHex:         tx.Hash().Hex(),
+		ContractAddress:   contractAddr.Hex(),
+		GasUsed:           res.UsedGas,
+		BlockNumber:       uint64(ctx.BlockHeight()),
+		TransactionIndex:  uint32(ctx.TxIndex()),
+		EffectiveGasPrice: tx.GasPrice().Uint64(),
+	}
+
+	if res.Err == nil {
+		receipt.Status = uint32(ethtypes.ReceiptStatusSuccessful)
+	} else {
+		receipt.Status = uint32(ethtypes.ReceiptStatusFailed)
+	}
+
+	return server.SetReceipt(ctx, tx.Hash(), receipt)
+}
+
+// returns a function that provides block header hash based on block number
+func (server msgServer) getHashFn(ctx sdk.Context) vm.GetHashFunc {
+	return func(height uint64) common.Hash {
+		if height > math.MaxInt64 {
+			ctx.Logger().Error("Sei block height is bounded by int64 range")
+			return common.Hash{}
+		}
+		h := int64(height)
+		if ctx.BlockHeight() == h {
+			// current header hash is in the context already
+			return common.BytesToHash(ctx.HeaderHash())
+		}
+		if ctx.BlockHeight() < h {
+			// future block doesn't have a hash yet
+			return common.Hash{}
+		}
+		// fetch historical hash from historical info
+		return server.getHistoricalHash(ctx, h)
+	}
+}
+
+func (server msgServer) getHistoricalHash(ctx sdk.Context, h int64) common.Hash {
```

Comment:
should we return error too? what is oss evm behavior?

Reply:
in ethereum it returns empty if block is not found https://github.com/ethereum/go-ethereum/blob/master/core/evm.go#L88